nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,2464 @@
+        info!(
+            log,
+            "instance update: successor update saga started!";
+            "instance_id" => %instance_id,
+        );
+    }
+
+    Ok(())
+}
+
+/// Unlock the instance record while unwinding.
+///
+/// This is factored out of the actual reverse action, because the `Params` type
+/// differs between the start saga and the actual instance update sagas, both of
+/// which must unlock the instance in their reverse actions.
+async fn unwind_instance_lock(
+    lock: instance::UpdaterLock,
+    serialized_authn: &authn::saga::Serialized,
+    authz_instance: &authz::Instance,
+    sagactx: &NexusActionContext,
+) {
+    // /!\ EXTREMELY IMPORTANT WARNING /!\
+    //
+    // This comment is a message, and part of a system of messages. Pay
+    // attention to it! The message is a warning about danger.
+    //
+    // The danger is still present in your time, as it was in ours. The danger
+    // is to the instance record, and it can deadlock.
+    //
+    // When unwinding, unlocking an instance MUST succeed at all costs. This is
+    // of the upmost importance. It's fine for unlocking an instance in a
+    // forward action to fail, since the reverse action will still unlock the
+    // instance when the saga is unwound. However, when unwinding, we must
+    // ensure the instance is unlocked, no matter what. This is because a
+    // failure to unlock the instance will leave the instance record in a
+    // PERMANENTLY LOCKED state, since no other update saga will ever be
+    // able to lock it again. If we can't unlock the instance here, our death
+    // will ruin the instance record forever and it will only be able to be
+    // removed by manual operator intervention. That would be...not great.
+    //
+    // Therefore, this action will retry the attempt to unlock the instance
+    // until it either:
+    //
+    // - succeeds, and we know the instance is now unlocked.
+    // - fails *because the instance doesn't exist*, in which case we can die
+    //   happily because it doesn't matter if the instance is actually unlocked.
+    use dropshot::HttpError;
+    use futures::{future, TryFutureExt};
+    use omicron_common::backoff;
+
+    let osagactx = sagactx.user_data();
+    let log = osagactx.log();
+    let instance_id = authz_instance.id();
+    let opctx =
+        crate::context::op_context_for_saga_action(sagactx, &serialized_authn);
+
+    debug!(
+        log,
+        "instance update: unlocking instance on unwind";
+        "instance_id" => %instance_id,
+        "lock" => ?lock,
+    );
+
+    const WARN_DURATION: std::time::Duration =
+        std::time::Duration::from_secs(20);
+
+    let did_unlock = backoff::retry_notify_ext(
+        // This is an internal service query to CockroachDB.
+        backoff::retry_policy_internal_service(),
+        || {
+            osagactx
+            .datastore()
+            .instance_updater_unlock(&opctx, authz_instance, &lock)
```

Comment:
No, a `time_deleted.is_null()` check was actually _removed_ from `instance_updater_unlock` in https://github.com/oxidecomputer/omicron/pull/5749/commits/c96adb62ea5b80af129563c361b4b6575bd54a3b, because it caused a race between an unwinding update saga and a (successful) instance-delete saga that caused the unwinding saga to get stuck when a stopped instance is deleted out from under it.

Reply:
OH, I see what you mean --- because `NotFound` doesn't make the unwind action fail, it's actually fine if deleted instances are filtered out, because we'll still just allow the saga to unwind successfully. Yeah, that makes sense.

Reply:
done in [591d51f](https://github.com/oxidecomputer/omicron/pull/5749/commits/591d51f500f036d380b9c1704882dd85dd3d5048)!